compair/algorithms/pair/adaptive_min_delta/pair_generator.py

```diff
@@ -0,0 +1,198 @@
+        - Second key is selected by minimum sum of weighted delta of each criteria
+          (if tied, select with closest score. if also tied, choose randomly)
+          with the lowest round possible
+        """
+        score_object_1 = None
+        score_object_2 = None
+
+        # step 1: select valid first element in pair
+        for round in self.rounds:
+            scored_objects = self.round_objects.get(round, [])
+
+            # place objects in round in random order
+            random.shuffle(scored_objects)
+
+            # find a scored_object that has a valid opponent
+            for index, scored_object in enumerate(scored_objects):
+                if self._has_valid_opponent(scored_object.key):
+                    score_object_1 = scored_object
+                    break
+
+            if score_object_1 != None:
+                break
+
+        if score_object_1 == None:
+            raise UserComparedAllObjectsException
+
+        # step 2: remove invalid opponents
+        self._remove_invalid_opponents(score_object_1.key)
+
+        # step 3: select valid second element in pair
+        """
+        select second element in pair
+        """
+        score = score_object_1.score
+        for round in self.rounds:
+            scored_objects = self.round_objects.get(round, [])
+
+            """
+            sort by followings, in order:
+            (1) sum of weighted criterion scores delta;
+            (2) absolute diff between the scores;
+            (3) a random number
+            """
+            scored_objects = sorted(scored_objects,
+                key=lambda scored_object: (
+                    self._criterion_score_delta_sum(score_object_1.key, scored_object.key),
+                    math.fabs(score - scored_object.score),
+                    random.random()
+                )
+            )
+
+            # take the first score object
+            if len(scored_objects) > 0:
+                score_object_2 = scored_objects[0]
+                break
+
+        if score_object_2 == None:
+            raise UnknownPairGeneratorException
+
+        return ComparisonPair(
+            key1=score_object_1.key,
+            key2=score_object_2.key,
+            winner=None
+        )
+
+    def _criterion_score_delta_sum(self, scored_object_key1, scored_object_key2):
+        """
+        Returns the sum of delta of criterion scores between the the scored obj
+        """
+        theSum = 0
+        criterion_scores_1 = self.criterion_scores.setdefault(scored_object_key1, {})
+        criterion_scores_2 = self.criterion_scores.setdefault(scored_object_key2, {})
+
+        criterion_key_list = list(criterion_scores_1.keys())
+        criterion_key_list.extend(list(criterion_scores_2.keys()))
+        criterion_key_list = set(criterion_key_list)
+
+        for criterion in criterion_key_list:
+            score1 = 0 if criterion_scores_1.setdefault(criterion, 0) == None \
```

Comment:
You should be able to write this to `score1 = criterion_scores_1.get(criterion, 0)` (second argument in dictionary get is default value)

same can be done for `score2` and `weight`

Reply:
@andrew-gardener agree that `get` is more suitable than `setdefault` here.  And the reason why I included checking of `None` was that during initial testing, I found the front end could pass in a dictionary of `{None : None}`.  Later I found I made a mistake by using outerjoin instead of join when retrieving data.  Now it shouldn't happen.

Will simply the logic and test again.